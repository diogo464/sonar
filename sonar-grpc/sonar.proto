syntax = "proto3";

package sonar;

import "google/protobuf/empty.proto";
import "google/protobuf/duration.proto";
import "google/protobuf/timestamp.proto";
import "google/protobuf/field_mask.proto";

service SonarService {
	rpc UserList(UserListRequest) returns (UserListResponse);
	rpc UserCreate(UserCreateRequest) returns (User);
	rpc UserUpdate(UserUpdateRequest) returns (User);
	rpc UserDelete(UserDeleteRequest) returns (google.protobuf.Empty);
	rpc UserLogin(UserLoginRequest) returns (UserLoginResponse);
	rpc UserLogout(UserLogoutRequest) returns (google.protobuf.Empty);

	rpc ImageCreate(ImageCreateRequest) returns (ImageCreateResponse);
	rpc ImageDelete(ImageDeleteRequest) returns (google.protobuf.Empty);
	rpc ImageDownload(ImageDownloadRequest) returns (stream ImageDownloadResponse);

	rpc ArtistList(ArtistListRequest) returns (ArtistListResponse);
	rpc ArtistGet(ArtistGetRequest) returns (Artist);
	rpc ArtistCreate(ArtistCreateRequest) returns (Artist);
	rpc ArtistUpdate(ArtistUpdateRequest) returns (Artist);
	rpc ArtistDelete(ArtistDeleteRequest) returns (google.protobuf.Empty);

	rpc AlbumList(AlbumListRequest) returns (AlbumListResponse);
	rpc AlbumListByArtist(AlbumListByArtistRequest) returns (AlbumListResponse);
	rpc AlbumGet(AlbumGetRequest) returns (Album);
	rpc AlbumCreate(AlbumCreateRequest) returns (Album);
	rpc AlbumUpdate(AlbumUpdateRequest) returns (Album);
	rpc AlbumDelete(AlbumDeleteRequest) returns (google.protobuf.Empty);

	rpc TrackList(TrackListRequest) returns (TrackListResponse);
	rpc TrackListByAlbum(TrackListByAlbumRequest) returns (TrackListResponse);
	rpc TrackGet(TrackGetRequest) returns (Track);
	rpc TrackCreate(TrackCreateRequest) returns (Track);
	rpc TrackUpdate(TrackUpdateRequest) returns (Track);
	rpc TrackDelete(TrackDeleteRequest) returns (google.protobuf.Empty);

	rpc PlaylistList(PlaylistListRequest) returns (PlaylistListResponse);
	rpc PlaylistGet(PlaylistGetRequest) returns (Playlist);
	rpc PlaylistCreate(PlaylistCreateRequest) returns (Playlist);
	rpc PlaylistUpdate(PlaylistUpdateRequest) returns (Playlist);
	rpc PlaylistDelete(PlaylistDeleteRequest) returns (google.protobuf.Empty);
	rpc PlaylistTrackList(PlaylistTrackListRequest) returns (PlaylistTrackListResponse);
	rpc PlaylistTrackInsert(PlaylistTrackInsertRequest) returns (google.protobuf.Empty);
	rpc PlaylistTrackRemove(PlaylistTrackRemoveRequest) returns (google.protobuf.Empty);
	rpc PlaylistTrackClear(PlaylistTrackClearRequest) returns (google.protobuf.Empty);

	rpc ScrobbleList(ScrobbleListRequest) returns (ScrobbleListResponse);
	rpc ScrobbleCreate(ScrobbleCreateRequest) returns (Scrobble);
	rpc ScrobbleDelete(ScrobbleDeleteRequest) returns (google.protobuf.Empty);

	rpc PinList(PinListRequest) returns (PinListResponse);
	rpc PinSet(PinSetRequest) returns (google.protobuf.Empty);
	rpc PinUnset(PinUnsetRequest) returns (google.protobuf.Empty);

	rpc ExternalDownloadStart(ExternalDownloadRequest) returns (google.protobuf.Empty);

	rpc Import(stream ImportRequest) returns (Track);

	rpc MetadataFetch(MetadataFetchRequest) returns (google.protobuf.Empty);

	rpc MetadataAlbumTracks(MetadataAlbumTracksRequest) returns (MetadataAlbumTracksResponse);
}

message Property {
	string key = 1;
	string value = 2;
}

message PropertyUpdate {
	string key = 1;
	optional string value = 2;
}

message User {
	string user_id = 1;
	string username = 2;
	optional string avatar_id = 3;
}

message UserListRequest {
	optional uint32 offset = 1;
	optional uint32 count = 2;
}

message UserListResponse {
	repeated User users = 1;
}

message UserCreateRequest {
	string username = 1;
	string password = 2;
	optional string avatar_id = 3;
}

message UserUpdateRequest {
	string user_id = 1;
	optional string password = 2;
	optional string avatar_id = 3;
}

message UserDeleteRequest {
	string user_id = 1;
}

message UserLoginRequest {
	string username = 1;
	string password = 2;
}

message UserLoginResponse {
	string user_id = 1;
	string token = 2;
}

message UserLogoutRequest {
	string token = 1;
}

message ImageCreateRequest {
	bytes content = 1;
}

message ImageCreateResponse {
	string image_id = 1;
}

message ImageDeleteRequest {
	string image_id = 1;
}

message ImageDownloadRequest {
	string image_id = 1;
}

message ImageDownloadResponse {
	string image_id = 1;
	string mime_type = 2;
	bytes content = 3;
}

message Artist {
	string id = 1;
	string name = 2;
	uint32 album_count = 3;
	uint32 listen_count = 4;
	optional string coverart_id = 5; 
	repeated Property properties = 6;
}

message ArtistListRequest {
	optional uint32 offset = 1;
	optional uint32 count = 2;
}

message ArtistListResponse {
	repeated Artist artists = 1;
}

message ArtistGetRequest {
	string artist_id = 1;
}

message ArtistCreateRequest {
	string name = 1;
	optional string coverart_id = 2;
	repeated Property properties = 3;
}

message ArtistDeleteRequest {
	string artist_id = 1;	
}

message ArtistUpdateRequest {
	string artist_id = 1;
	optional string name = 2;
	optional string coverart_id = 3;
	repeated PropertyUpdate properties = 4;
}

message Album {
	string id = 1;
	string name = 2;
	uint32 track_count = 3;
	google.protobuf.Duration duration = 4;
	uint32 listen_count = 5;
	string artist_id = 6;
	optional string coverart_id = 7;
	repeated Property properties = 8;
}

message AlbumListRequest {
	optional uint32 offset = 1;
	optional uint32 count = 2;
}

message AlbumListByArtistRequest {
	optional uint32 offset = 1;
	optional uint32 count = 2;
	string artist_id = 3;
}

message AlbumListResponse {
	repeated Album albums = 1;
}

message AlbumGetRequest {
	string album_id = 1;
}

message AlbumCreateRequest {
	string name = 1;
	string artist_id = 2;
	optional string coverart_id = 3;
	repeated Property properties = 4;
}

message AlbumUpdateRequest {
	string album_id = 1;
	optional string name = 2;
	optional string artist_id = 3;
	optional string coverart_id = 4;
	repeated PropertyUpdate properties = 5;
}

message AlbumDeleteRequest {
	string album_id = 1;
}

message Track {
	string id = 1;
	string name = 2;
	string artist_id = 3;
	string album_id = 4;
	google.protobuf.Duration duration = 5;
	uint32 listen_count = 6;
	optional string cover_art_id = 7;
	repeated Property properties = 8;
}

message TrackListRequest {
	optional uint32 offset = 1;
	optional uint32 count = 2;
}

message TrackListByAlbumRequest {
	optional uint32 offset = 1;
	optional uint32 count = 2;
	string album_id = 3;
}

message TrackListResponse {
	repeated Track tracks = 1;
}

message TrackGetRequest {
	string track_id = 1;
}

message TrackCreateRequest {
	string name = 1;
	string album_id = 2;
	optional string coverart_id = 3;
	optional string audio_id = 4;
	repeated Property properties = 5;
}

message TrackUpdateRequest {
	string track_id = 1;
	optional string name = 2;
	optional string album_id = 3;
	optional string coverart_id = 4;
	repeated PropertyUpdate properties = 5;
}

message TrackDeleteRequest {
	string track_id = 1;
}

message Playlist {
	string id = 1;
	string name = 2;
	string user_id = 3;
	uint32 track_count = 4;
	google.protobuf.Duration duration = 5;
	optional string coverart_id = 6;
	repeated Property properties = 7;
}

message PlaylistListRequest {
	optional uint32 offset = 1;
	optional uint32 count = 2;
	optional string user_id = 3;
}

message PlaylistListResponse {
	repeated Playlist playlists = 1;
}

message PlaylistGetRequest {
	string playlist_id = 1;
}

message PlaylistCreateRequest {
	string name = 1;
	string owner_id = 2;
	repeated string track_ids = 3;
	repeated Property properties = 4;
}

message PlaylistUpdateRequest {
	string playlist_id = 1;
	optional string name = 2;
	repeated PropertyUpdate properties = 3;
}

message PlaylistDeleteRequest {
	string playlist_id = 1;
}

message PlaylistTrackListRequest {
	string playlist_id = 1;
}

message PlaylistTrackListResponse {
	repeated Track tracks = 1;
}

message PlaylistTrackInsertRequest {
	string playlist_id = 1;
	repeated string track_ids = 2;
}

message PlaylistTrackRemoveRequest {
	string playlist_id = 1;
	repeated string track_ids = 2;
}

message PlaylistTrackClearRequest {
	string playlist_id = 1;
}

message Scrobble {
	string id = 1;
	string track_id = 2;
	string user_id = 3;
	google.protobuf.Timestamp listen_at = 4;
	google.protobuf.Duration listen_duration = 5;
	string listen_device = 6;
	repeated Property properties = 7;
}

message ScrobbleListRequest {
	optional uint32 offset = 1;
	optional uint32 count = 2;
}

message ScrobbleListResponse {
	repeated Scrobble scrobbles = 1;
}

message ScrobbleCreateRequest {
	string user_id = 1;
	string track_id = 2;
	google.protobuf.Timestamp listen_at = 3;
	google.protobuf.Duration listen_duration = 4;
	string listen_device = 5;
	repeated Property properties = 6;
}

message ScrobbleDeleteRequest {
	string scrobble_id = 1;
}

message PinListRequest {
	string user_id = 1;
}

message PinListResponse {
	repeated string sonar_ids = 1;
}

message PinSetRequest {
	string user_id = 1;
	repeated string sonar_ids = 2;
}

message PinUnsetRequest {
	string user_id = 1;
	repeated string sonar_ids = 2;
}

message ExternalDownloadRequest {
	string user_id = 1;
	string external_id = 2;
}

message ImportRequest {
	bytes chunk = 1;
	optional string filepath = 2;
	optional uint32 artist_id = 3;
	optional uint32 album_id = 4;
}

enum MetadataFetchKind {
	METADATA_FETCH_KIND_ARTIST = 0;
	METADATA_FETCH_KIND_ALBUM = 1;
	METADATA_FETCH_KIND_ALBUMTRACKS = 2;
	METADATA_FETCH_KIND_TRACK = 3;
}

message MetadataFetchRequest {
	MetadataFetchKind kind = 1;
	string item_id = 2;
}

message TrackMetadata {
	optional string name = 1;
	repeated Property properties = 2;
	optional bytes cover = 3;
}

message MetadataAlbumTracksRequest {
	uint32 album_id = 1;
}

message MetadataAlbumTracksResponse {
	map<uint32, TrackMetadata> tracks = 1;
}

message SpotifyListRequest {
}

message SpotifyListResponse {
	repeated string spotify_ids = 1;
}

message SpotifyAddRequest {
	repeated string spotify_ids = 1;
}

message SpotifyRemoveRequest {
	repeated string spotify_ids = 1;
}
