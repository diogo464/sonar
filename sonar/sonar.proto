syntax = "proto3";

package sonar;

import "google/protobuf/empty.proto";
import "google/protobuf/field_mask.proto";

service SonarService {
	// rpc UserList(UserListRequest) returns (UserListResponse);
	// rpc UserCreate(UserCreateRequest) returns (UserCreateResponse);
	// rpc UserDelete(UserDeleteRequest) returns (UserDeleteResponse);

	rpc ArtistList(ArtistListRequest) returns (ArtistListResponse);
	rpc ArtistCreate(ArtistCreateRequest) returns (Artist);
	rpc ArtistUpdate(ArtistUpdateRequest) returns (Artist);
	rpc ArtistDelete(ArtistDeleteRequest) returns (google.protobuf.Empty);
}

// A positive duration value
message Duration {
	uint32 nanoseconds = 1;
}

// A unix timestamp
message Timestamp {
	uint64 seconds = 1;
	uint32 nanoseconds = 2;
}

message Property {
	string key = 1;
	string value = 2;
}

message PropertyUpdate {
	string key = 1;
	optional string value = 2;
}

message GenresUpdate {
}

message User {
	uint32 id = 1;
	string username = 2;
	optional uint32 avatar = 3;
}

message Artist {
	uint32 id = 1;
	string name = 2;
	uint32 album_count = 3;
	uint32 listen_count = 4;
	repeated string genres = 5;
	optional uint32 coverart = 6; 
	repeated Property properties = 7;
}

message ArtistListRequest {
	optional uint32 offset = 1;
	optional uint32 count = 2;
}

message ArtistListResponse {
	repeated Artist artists = 1;
}

message ArtistCreateRequest {
	string name = 1;
	repeated string genres = 2;
	optional uint32 coverart = 3;
	repeated Property properties = 4;
}

message ArtistDeleteRequest {
	uint32 artist = 1;	
}

message ArtistUpdateRequest {
	uint32 artist = 1;
	optional string name = 2;
	repeated PropertyUpdate properties = 4;
	google.protobuf.FieldMask mask = 5;
}

message Album {
	uint32 id = 1;
	string name = 2;
	uint32 track_count = 3;
	Duration duration = 4;
	uint32 listen_count = 5;
	repeated uint32 artists = 6;
	repeated string genres = 7;
	optional uint32 coverart = 8;
	repeated Property properties = 9;
}

message Track {
	uint32 id = 1;
	string name = 2;
	Duration duration = 3;
	uint32 album = 4;
	uint32 listen_count = 5;
	repeated uint32 artists = 6;
	optional uint32 coverart = 7;
	repeated Property properties = 8;
}

message Playlist {
	uint32 id = 1;
	string name = 2;
	uint32 user = 3;
	uint32 track_count = 4;
	Duration duration = 5;
	optional uint32 coverart = 6;
	repeated Property properties = 7;
}

message Scrobble {
	uint32 id = 1;
	uint32 track = 2;
	uint32 user = 3;
	Duration duration = 4;
	Timestamp timestamp = 5;
	repeated Property properties = 6;
}
