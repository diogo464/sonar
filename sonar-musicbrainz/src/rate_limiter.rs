use std::{
    sync::Arc,
    time::{Duration, Instant},
};

use tokio::sync::Mutex;

#[derive(Debug, Clone)]
pub struct RateLimiter(Arc<Mutex<Inner>>);

#[derive(Debug)]
struct Inner {
    rps: f32,
    last_req: Option<Instant>,
}

impl RateLimiter {
    pub fn new(rps: f32) -> Self {
        Self(Arc::new(Mutex::new(Inner {
            rps,
            last_req: None,
        })))
    }

    pub async fn request(&self) {
        let mut inner = self.0.lock().await;
        if let Some(elapsed) = inner.last_req.map(|i| i.elapsed()) {
            let interval = Duration::from_secs_f32(1.0 / inner.rps);
            if elapsed < interval {
                let sleep = interval - elapsed;
                tokio::time::sleep(sleep).await;
            }
        }
        inner.last_req = Some(Instant::now());
    }
}
