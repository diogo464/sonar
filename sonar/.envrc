export DATABASE_URL=sqlite:/home/diogo464/dev/code/sonar/sonar/sonar.db
