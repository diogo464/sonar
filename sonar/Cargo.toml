[package]
name = "sonar"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[features]
test-utilities = ["dep:rand"]

[dependencies]
async-trait = "0.1.77"
bytes = "1.5.0"
chrono = "0.4.33"
scrypt = "0.11.0"
sqlx = { version = "0.7.3", features = ["sqlite", "runtime-tokio"] }
tempfile = "3.9.0"
tokio = { version = "1.35.1", features = ["full"] }
tokio-stream = { version = "0.1.14", features = ["full"] }
tokio-util = { version = "0.7.10", features = ["full"] }
tracing = "0.1.40"
ulid = "1.1.0"

rand = { version = "0.8.5", optional = true }
serde = { version = "1.0.196", features = ["derive"] }
bincode = "1.3.3"
infer = "0.15.0"
lofty = "0.18.2"
sha2 = "0.10.8"
hex = "0.4.3"

[dev-dependencies]
tracing-subscriber = { version = "0.3.18", features = ["env-filter"] }
sonar = { path = "." , features = ["test-utilities"] }
