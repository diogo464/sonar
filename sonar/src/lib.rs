#![feature(const_for)]
#![feature(const_trait_impl)]
#![feature(concat_idents)]
#![feature(backtrace_frames)]
#![feature(let_chains)]
#![feature(is_sorted)]

mod error;
pub use error::*;

mod id;
pub use id::*;

mod types;
pub use types::*;

mod context;
pub use context::*;

pub mod bytestream;
pub mod ext;
pub mod prop;

#[doc(hidden)]
#[cfg(feature = "test-utilities")]
pub mod test;

pub(crate) mod album;
pub(crate) mod artist;
pub(crate) mod audio;
pub(crate) mod blob;
pub(crate) mod db;
pub(crate) mod download;
pub(crate) mod external;
pub(crate) mod extractor;
pub(crate) mod gc;
pub(crate) mod genre;
pub(crate) mod image;
pub(crate) mod importer;
pub(crate) mod ks;
pub(crate) mod metadata;
pub(crate) mod pin;
pub(crate) mod playlist;
pub(crate) mod property;
pub(crate) mod scrobble;
pub(crate) mod scrobbler;
pub(crate) mod search;
pub(crate) mod subscription;
pub(crate) mod track;
pub(crate) mod user;

pub use album::{Album, AlbumCreate, AlbumUpdate};
pub use artist::{Artist, ArtistCreate, ArtistUpdate};
pub use audio::{Audio, AudioCreate, AudioDownload};
pub use download::{Download, DownloadCreate, DownloadDelete, DownloadStatus};
pub use external::{
    ExternalAlbum, ExternalArtist, ExternalImage, ExternalMediaId, ExternalMediaType,
    ExternalPlaylist, ExternalService, ExternalTrack,
};
pub use extractor::{ExtractedImage, ExtractedMetadata, Extractor};
pub use genre::{Genre, GenreUpdateAction, Genres, InvalidGenreError};
pub use image::{ImageCreate, ImageDownload};
pub use importer::Import;
pub use metadata::{
    AlbumMetadata, AlbumMetadataRequest, AlbumTracksMetadata, AlbumTracksMetadataRequest,
    ArtistMetadata, ArtistMetadataRequest, MetadataProvider, MetadataRequestKind, TrackMetadata,
    TrackMetadataRequest,
};
pub use playlist::{Playlist, PlaylistCreate, PlaylistTrack, PlaylistUpdate};
pub use property::{
    InvalidPropertyKeyError, InvalidPropertyValueError, Properties, PropertyKey, PropertyUpdate,
    PropertyUpdateAction, PropertyValue,
};
pub use scrobble::{Scrobble, ScrobbleCreate, ScrobbleUpdate};
pub use scrobbler::Scrobbler;
pub use search::{SearchFlags, SearchQuery, SearchResult};
pub use subscription::{Subscription, SubscriptionCreate, SubscriptionDelete};
pub use track::{Lyrics, LyricsKind, Track, TrackCreate, TrackLyrics, TrackUpdate};
pub use user::{
    InvalidUserTokenError, InvalidUsernameError, User, UserCreate, UserToken, UserUpdate, Username,
};

pub use async_trait::async_trait;
pub use bytes;
pub use chrono;

pub mod metadata_prelude {
    pub use crate::{
        Album, AlbumId, AlbumMetadata, AlbumMetadataRequest, AlbumTracksMetadata,
        AlbumTracksMetadataRequest, Artist, ArtistId, ArtistMetadata, ArtistMetadataRequest, Error,
        ErrorKind, MetadataProvider, MetadataRequestKind, Properties, Result, Track, TrackId,
        TrackMetadata, TrackMetadataRequest,
    };
}
