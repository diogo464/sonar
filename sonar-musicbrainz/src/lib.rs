use serde::Deserialize;
use sonar::{
    bytestream::ByteStream, Error, ErrorKind, ExternalAlbum, ExternalArtist,
    ExternalMediaEnrichStatus, ExternalMediaId, ExternalMediaRequest, ExternalMediaType,
    ExternalPlaylist, ExternalTrack, MultiExternalMediaId, Result,
};

mod rate_limiter;
use rate_limiter::*;

const USER_AGENT: &str = "sonar/1.0.0 ( diogo464@d464.sh )";

#[derive(Debug)]
pub struct MusicBrainzService {
    client: reqwest::Client,
    limiter: RateLimiter,
}

impl Default for MusicBrainzService {
    fn default() -> Self {
        Self {
            client: Default::default(),
            limiter: RateLimiter::new(1.0),
        }
    }
}

impl MusicBrainzService {
    async fn expand_recording(&self, ids: MultiExternalMediaId) -> MultiExternalMediaId {
        #[derive(Debug, Deserialize)]
        struct Response {
            relations: Vec<ResponseRelation>,
        }
        #[derive(Debug, Deserialize)]
        struct ResponseRelation {
            url: Option<ResponseRelationUrl>,
        }
        #[derive(Debug, Deserialize)]
        struct ResponseRelationUrl {
            resource: String,
        }

        let mut external_ids = Vec::with_capacity(ids.len());
        for id in ids {
            external_ids.push(id.clone());
            if !id_is_mbid(id.as_str()) {
                continue;
            }

            tracing::debug!("waiting for rate limiter");
            self.limiter.request().await;

            tracing::debug!("looking up recording: {id}");
            let response = self
                .client
                .get(format!(
                    "https://musicbrainz.org/ws/2/recording/{}?inc=url-rels",
                    id.as_str()
                ))
                .header("Accept", "application/json")
                .header("User-Agent", USER_AGENT)
                .send()
                .await
                .unwrap();

            if response.status() == reqwest::StatusCode::NOT_FOUND {
                continue;
            }

            let content = response.text().await.unwrap();
            tracing::debug!("recording response: {content}");
            let response: Response = serde_json::from_str(&content).unwrap();
            for relation in response.relations {
                if let Some(url) = relation.url {
                    external_ids.push(ExternalMediaId::from(url.resource));
                }
            }
        }
        MultiExternalMediaId::from(external_ids)
    }
}

#[sonar::async_trait]
impl sonar::ExternalService for MusicBrainzService {
    #[tracing::instrument(skip(self))]
    async fn enrich(
        &self,
        request: &mut ExternalMediaRequest,
    ) -> Result<ExternalMediaEnrichStatus> {
        Ok(ExternalMediaEnrichStatus::NotModified)
    }
    #[tracing::instrument(skip(self))]
    async fn extract(
        &self,
        request: &ExternalMediaRequest,
    ) -> Result<(ExternalMediaType, ExternalMediaId)> {
        Err(Error::new(ErrorKind::Internal, "not implemented"))
    }
    async fn fetch_artist(&self, _id: &ExternalMediaId) -> Result<ExternalArtist> {
        Err(Error::new(
            ErrorKind::Invalid,
            "musicbrainz service does not support fetching artists",
        ))
    }
    async fn fetch_album(&self, _id: &ExternalMediaId) -> Result<ExternalAlbum> {
        Err(Error::new(
            ErrorKind::Invalid,
            "musicbrainz service does not support fetching albums",
        ))
    }
    async fn fetch_track(&self, _id: &ExternalMediaId) -> Result<ExternalTrack> {
        Err(Error::new(
            ErrorKind::Invalid,
            "musicbrainz service does not support fetching tracks",
        ))
    }
    async fn fetch_playlist(&self, _id: &ExternalMediaId) -> Result<ExternalPlaylist> {
        Err(Error::new(
            ErrorKind::Invalid,
            "musicbrainz service does not support fetching playlists",
        ))
    }
    async fn download_track(&self, _id: &ExternalMediaId) -> Result<ByteStream> {
        Err(Error::new(
            ErrorKind::Invalid,
            "musicbrainz service does not support downloading tracks",
        ))
    }
}

fn id_is_mbid(id: &str) -> bool {
    let id = id.as_bytes();
    if id.len() != 36 {
        return false;
    }
    if id[8] != id[13] || id[13] != id[18] || id[18] != id[23] || id[23] != b'-' {
        return false;
    }
    id.iter()
        .all(|c| c.is_ascii_digit() || (*c >= b'a' && *c <= b'f') || *c == b'-')
}

#[cfg(test)]
mod test {
    use super::*;

    #[tokio::test]
    async fn test_expand() {
        let service = MusicBrainzService::default();
        let recording_mbid = ExternalMediaId::from("307ce9da-5690-4e21-ab71-9d12ea106e52");
        let expanded = service
            .expand_recording(MultiExternalMediaId::from(recording_mbid))
            .await;
        insta::assert_debug_snapshot!(expanded);
    }

    #[test]
    fn test_id_is_mbid() {
        assert!(id_is_mbid("307ce9da-5690-4e21-ab71-9d12ea106e52"));
        assert!(!id_is_mbid("spotify:artist:762310PdDnwsDxAQxzQkfX"));
    }
}
