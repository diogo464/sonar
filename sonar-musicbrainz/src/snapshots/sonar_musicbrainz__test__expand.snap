---
source: sonar-musicbrainz/src/lib.rs
expression: expanded
---
MultiExternalMediaId(
    [
        ExternalMediaId(
            "307ce9da-5690-4e21-ab71-9d12ea106e52",
        ),
        ExternalMediaId(
            "https://open.spotify.com/track/1mea3bSkSGXuIRvnydlB5b",
        ),
        ExternalMediaId(
            "https://www.deezer.com/track/3157972",
        ),
    ],
)
